src/operator/nn/mkldnn/mkldnn_softmax.cc

```diff
@@ -21,34 +21,109 @@
  * \file mkldnn_softmax.cc
  * \brief
  * \author Da Zheng
+ * \author Wenting Jiang (wenting.jiang@intel.com)        
 */
 
+#if MXNET_USE_MKLDNN == 1
 #include "../softmax-inl.h"
 #include "./mkldnn_ops-inl.h"
 #include "./mkldnn_base-inl.h"
 
-#if MXNET_USE_MKLDNN == 1
 namespace mxnet {
 namespace op {
 
-void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
-                          const NDArray &in_data, const OpReqType &req,
+class MKLDNNSoftmaxFwd {
+  std::shared_ptr<mkldnn::softmax_forward> fwd;
+  std::shared_ptr<mkldnn::memory> data;
+  std::shared_ptr<mkldnn::memory> out;
+
+ public:
+  MKLDNNSoftmaxFwd(const SoftmaxParam& param,
+                   bool is_train,
+                   const NDArray &in_data,
+                   const NDArray &out_data,
+                   const OpReqType &req):
+                   fwd(nullptr), data(nullptr), out(nullptr) {
+                   _Init(param, is_train, in_data, out_data, req);
+  }
+  ~MKLDNNSoftmaxFwd() {}
+  void SetDataHandle(const NDArray &in_data,
+                     const NDArray &out_data,
+                     const OpReqType &req) {
+    this->data->set_data_handle(in_data.GetMKLDNNData()->get_data_handle());
+    auto out_mem = const_cast<NDArray&>(out_data).CreateMKLDNNData(this->out->get_primitive_desc());
+    this->out->set_data_handle(out_mem->get_data_handle());
+  }
+  void Execute() {
+    MKLDNNStream *stream = MKLDNNStream::Get();
+    stream->RegisterPrim(*fwd);
+    stream->Submit();
+  }
+
+ private:
+  void _Init(const SoftmaxParam& param,
+             bool is_train,
+             const NDArray &in_data,
+             const NDArray &out_data,
+             const OpReqType &req) {
+    // mkldnn::softmax_forward::primitive_desc
+    auto input_mem = in_data.GetMKLDNNData();
+    auto output_mem = out_data.GetMKLDNNData();
```

Comment:
Calling GetMKLDNNData() will create memory for out_data with default layout. But out_data may need a specific layout which is same as in_data?